### PR TITLE
feat: add --use-most-recent cache for bulk upsert and delete

### PR DESCRIPTION
**Pull request by peternhale**
### What does this PR do?

- Restores force:data commands

1. force:data:bulk:upsert
2. force:data:bulk:upsert
3. force:data:bulk:status
- Deprecates command `data:resume`

- Add resume commands
1. data:upsert:resume
2. data:delete:resume

- Modified bulk upsert and delete 

1. Use Bulk API V2
2. Added TTL cache for async
3. Added async flags

### What issues does this PR fix or reference?
@W-12083162@

### CODE REVIEW

File: src/bulkDataRequestCache.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    };
+  }
+
+  /**
+   * Creates a new bulk data request cache entry for the given bulk request id.
+   *
+   * @param bulkRequestId
+   * @param username
+   */
+  public async createCacheEntryForRequest(bulkRequestId: string, username: Optional<string>, apiVersion: string | undefined): Promise<void> {
```
Comment (WillieRuemmele) [inline src/bulkDataRequestCache.ts:40]: NIT: consistently type these, `Optional<string>`=`string | undefined`, or use `?`

File: src/bulkDataRequestCache.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+  public async resolveResumeOptionsFromCache(
+    bulkQueryId: string | undefined,
+    useMostRecent: boolean,
+    org: Org | undefined,
+    apiVersion: string | undefined): Promise<ResumeOptions> {
+    if (!useMostRecent && !bulkQueryId) {
+      throw messages.createError('bulkRequestIdRequiredWhenNotUsingMostRecent');
+    }
+    const resumeOptions: ResumeOptions = {
```
Comment (WillieRuemmele) [inline src/bulkDataRequestCache.ts:61]: are all these typings required? `: ResumeOptions` and `as ResumeOptions`?

File: src/commands/data/upsert/bulk.ts [hunk 3]
```diff
@@ -4,88 +4,47 @@
...
-      char: 'r',
-      summary: messages.getMessage('flags.serial'),
-      default: false,
-    }),
+      deprecateAliases: true
+    })
   };
 
-  public async run(): Promise<BatcherReturnType> {
+  public async run(): Promise<BulkResultV2> {
```
Comment (WillieRuemmele) [inline src/commands/data/upsert/bulk.ts:33]: Does `BatcherReturnType===BulkResultV2`?

File: src/reporters.ts [hunk 4]
```diff
@@ -375,3 +377,60 @@ export const escape = (value: string): string => {
...
+
+export const BatchInfoColumns = {
+  id: { header: 'Batch Id' },
+  state: { header: 'State' },
+  failed: { header: 'Failed' },
+  stateMessage: { header: 'Message' },
+};
+
+export const getBatchTotals = (batches: BatchInfo[]): { total: number; failed: number; success: number } => {
+  const ttls = batches.reduce(
```
Comment (WillieRuemmele) [inline src/reporters.ts:389]: `return batches.reduce...`?

File: src/reporters.ts [hunk 5]
```diff
@@ -375,3 +377,60 @@ export const escape = (value: string): string => {
...
+  ttls.failed = Object.keys(results.failedResults).length;
+  ttls.success = Object.keys(results.successfulResults).length;
+  ttls.unprocessed = Object.keys(results.unprocessedRecords).length;
+  return ttls;
+}
+
+export const getFailedBatchesForDisplay = (
+  batches: BatchInfo[]
+): Array<{ id: string; state: BatchState; failed: string; stateMessage: string }> => {
+  const failedBatches = batches.filter((batch) => parseInt(batch.numberRecordsFailed, 10) > 0);
```
Comment (WillieRuemmele) [inline src/reporters.ts:418]: can you combine this filter/parseInt with the map below, and just return that without creating a new constant?

**Comment by WillieRuemmele** [comment]
### QA Notes:
---

### `force:data:bulk:upsert`
---
`-h` : ✅ 
suggests `data resume` command ✅ 
deprecated flags warnings ✅ 
`--wait` ✅ 

### `force:data:bulk:status`
---
❌ : suggests itself as an alternative?
```
 ➜  sfdx force:data:bulk:status -i 750DR0000062iNMYAY -b 751DR000006oVClYAM
Warning: The "force data bulk status" command has been deprecated. Use "data resume" instead.
Warning: Use force:data:bulk:status instead
```

maybe it's an issue with `data resume`?
```
 ➜  sfdx data resume -i 750DR0000062iNMYAY -b 751DR000006oVClYAM                                           
Warning: Use force:data:bulk:status instead
```

### `force:data:bulk:delete`
---
💡 : we could improve the error when an invalid sobject is passed 
```
 ➜  sfdx force:data:bulk:delete --file accounts.csv --sobject=Accounts
Bulk Delete... Error
InvalidJob: Unable to find object: Accounts
    at BulkApi.getError (/Users/william.ruemmele/projects/oss/plugin-data/node_modules/jsforce/lib/http-api.js:307:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /Users/william.ruemmele/projects/oss/plugin-data/node_modules/jsforce/lib/http-api.js:154:23
    at async /Users/william.ruemmele/projects/oss/plugin-data/node_modules/jsforce/lib/api/bulk.js:163:23
    at async PassThrough.<anonymous> (/Users/william.ruemmele/projects/oss/plugin-data/node_modules/jsforce/lib/api/bulk.js:419:9) {
  errorCode: 'InvalidJob',
  content: undefined
}
```

### `data upsert bulk`
---
✅ : all looks good 
- json
- help
- human
- wait
- timeout

❌ 
```
Run command sfdx data upsert resume -i 750DR0000062iQLYAY -o test-vs9y4m8v7jq9@example.com' to check status.
```
there's an extra `'` at the end of the username which doesn't match anything

❓ 
the `force:bulk:upsert --wait` commands will print a large summary table, generally I'm unsure how to get this same output from the new commands, there's no `--wait` on the `resume` command. Adding `--wait` to `data bulk upsert` just results in printing 
```
 ➜  sfdx data bulk upsert --file accounts.csv --sobject Account --external-id Id --wait 3   

Job 750DR0000062iQuYAI Status Job Complete Records processed 4. Records failed 0.
Running bulk upsert request... done
``` 

### `data upsert resume`
---
✅ : all looks good
- help
- human
- cache
- passed id
- no cache interference with `data delete resume`

### `data bulk delete`
✅ : all looks good, but same trailing `'` in the output
❌  : failed delete results in an invalid command?

```
 ➜  sfdx data delete resume
Getting status... done

Job 750DR0000062iTAYAY Status Failed Records processed 0. Records failed 0.
To review the details of this job, run:
sfdx org open --target-org test-vs9y4m8v7jq9@example.com --path "/lightning/setup/AsyncApiJobStatus/page?address=%2F750DR0000062iTAYAY"

 ➜  sfdx org open --target-org test-vs9y4m8v7jq9@example.com --path "/lightning/setup/AsyncApiJobStatus/page?address=%2F750DR0000062iTAYAY"

 ›   Warning: org open is not a sfdx command.
Did you mean version? [y/n]: 
 ›   Error: Run sfdx help for a list of available commands.
➜  dreamhouse-lwc git:(main) ✗  hub:(GLOBAL - DevHub) scratch:(test-vs9y4m8v7jq9@example.com)
 ➜  sf org open --target-org test-vs9y4m8v7jq9@example.com --path "/lightning/setup/AsyncApiJobStatus/page?address=%2F750DR0000062iTAYAY" 

 ›   Warning: @salesforce/cli update available from 1.61.1 to 1.62.2.
 ›   Warning: org open is not a sf command.
Did you mean env open? [y/n]: n
 ›   Error: Run sf help for a list of available commands.
```

**Comment by WillieRuemmele** [comment]
✅ : circular deprecation fixed
❓ : still unsure how to get something similar to 
```
 ➜  sfdx force:data:bulk:upsert --file accounts.csv --sobject Account --external-id Id -w 3
...

=== Batch #1

=== Batch Status

apexProcessingTime:      0
apiActiveProcessingTime: 114
```
❌ : still suggesting invalid `org open` command - maybe fine once plugin-org is sf-ified?
🆗 : the extra `'` is still present

**Review by WillieRuemmele — APPROVED** [review]

File: src/bulkUtils.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+export const validateSobjectType = async (sobjectType: string, connection: Connection): Promise<void> => {
+  try {
+    await connection.sobject(sobjectType).describe();
+  } catch (e) {
+    throw new Error(messages.getMessage('invalidSobject', [sobjectType, (e as Error).message]));
+  }
+};
+
+export const waitOrTimeout = async (job: IngestJobV2<Schema, IngestOperation>, wait: number): Promise<void> => {
```
Comment (shetzel) [inline src/bulkUtils.ts:36]: Why not use core's PollingClient?

File: test/commands/data/dataBulk.nut.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+      if (result.jsonOutput.result.jobInfo.state === 'JobComplete') {
+        complete = true;
+      } else {
+        // eslint-disable-next-line no-console
+        console.log(result);
+      }
+    } else {
+      // eslint-disable-next-line no-console
+      console.log(result);
```
Comment (shetzel) [inline test/commands/data/dataBulk.nut.ts:38]: Is this debugging?  logging to the console during tests can really clutter test output.

**Review by shetzel — APPROVED** [review]

**Comment by WillieRuemmele** [comment]
### QA
---

❌ : undefined printed to user
```
 ➜  sfdx data delete resume -i 750DS000007vdNBYAY -o test-tqzt9pi0rxxn@example.com
Getting status... done

Job 750DS000007vdNBYAY Status Failed Records processed 0. Records failed 0.
To review the details of this job, run:
sfdx org open --target-org undefined --path "/lightning/setup/AsyncApiJobStatus/page?address=%2F750DS000007vdNBYAY"
```
✅ : QA'd and verified a lot of the other behaviors marked as ❌  in previous QA comments and the rest of the commands

File: test/commands/data/dataBulk.nut.ts (hunk 7 above)
Comment (peternhale) [reply test/commands/data/dataBulk.nut.ts:38]: Removed

**Pull request merged**